### PR TITLE
Speakers: Add seperate resourcelist class, docs, permissions for dire…

### LINKED ISSUE #4148: Speakers: Add seperate resourcelist class, docs, permissions for direct endpoints
Parent issue #4101.

### PR DESCRIPTION
**Pull request by enigmaeth**
…ct endpoints

Fixes #4148.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by enigmaeth** [comment]
@fossasia/open-event-server-devs Please review.

**Comment by shubham-padia** [comment]
@enigmaeth Points no. 2 and 4 from https://github.com/fossasia/open-event/blob/master/dev-handbook/resources-permissions.md#speakers---eg-v1speakers seem to be missing

**Comment by enigmaeth** [comment]
@shubham-padia Updated. Please check now!

File: app/api/events.py [hunk 1]
```diff
@@ -104,7 +105,7 @@ def validate_timezone(self, data, original_data):
     organizer_description = fields.Str(allow_none=True)
     is_sessions_speakers_enabled = fields.Bool(default=False)
     privacy = fields.Str(default="public")
-    state = fields.Str(default="Draft")
+    state = fields.Str(validate=validate.OneOf(choices=["published", "draft"]), allow_none=True, default='draft')
```
Comment (shubham-padia) [inline app/api/events.py]: Here, published and draft are not the only acceptable values. Please populate the above validator with all possible values.
While,
> Only of sessions with state approved or accepted

means is speaker with approved or accepted sessions should be there when a `GET` request is done. For this you'd have to modify the query method here I think

**Comment by enigmaeth** [comment]
@shubham-padia Added migrations. Please review!

File: migrations/versions/4defe72abed3_.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('sessions', sa.Column('creator_id', sa.Integer(), nullable=True))
+    op.create_foreign_key(None, 'sessions', 'users', ['creator_id'], ['id'], ondelete='CASCADE')
+    op.add_column('sessions_version', sa.Column('creator_id', sa.Integer(), autoincrement=False, nullable=True))
+    op.execute("""UPDATE events SET state = LOWER(state)""")
+    op.execute("""UPDATE events SET state = 'draft' WHERE state != 'published'""")
+    op.execute("""UPDATE sessions SET state = LOWER(state)""")
+    op.execute("""UPDATE sessions SET state = 'draft' WHERE state not in 
+        ('accepted', 'pending', 'approved', 'rejected')""")
```
Comment (enigmaeth) [inline migrations/versions/4defe72abed3_.py]: I think we don't actually need downgrade statements for these.

**Comment by enigmaeth** [comment]
@shubham-padia @niranjan94 Can you review this now? Added the migrations!

**Comment by enigmaeth** [comment]
@shubham-padia @niranjan94 Please review.

**Comment by enigmaeth** [comment]
@niranjan94 @shubham-padia 
Added docs and dredd tests for GET 
- `events/id/speakers`
- `users/id/speakers`
- `sessions/id/speakers`
Please review again :D

File: app/api/speakers.py [hunk 3]
```diff
@@ -111,42 +92,68 @@ def before_post(self, args, kwargs, data):
         :return:
         """
         require_relationship(['event', 'user'], data)
-        kwargs['user_id'] = current_identity.id
 
-    def before_create_object(self, data, view_kwargs):
+        if not has_access('is_coorganizer', event_id=data['event']):
+            if has_access('is_user_itself'):
```
Comment (shubham-padia) [inline app/api/speakers.py]: I'm not sure if `is_user_itself` can be used here, and it checks `kwargs['id'] == user.id`, can you confirm please.

Comment (enigmaeth) [inline app/api/speakers.py]: @shubham-padia  Yeah, you are right. Updated now, please check.

File: app/api/speakers.py [hunk 4]
```diff
@@ -111,42 +92,68 @@ def before_post(self, args, kwargs, data):
...
         """
         require_relationship(['event', 'user'], data)
-        kwargs['user_id'] = current_identity.id
 
-    def before_create_object(self, data, view_kwargs):
+        if not has_access('is_coorganizer', event_id=data['event']):
+            if has_access('auth_required'):
+                event = safe_query(self, Event, 'id', data['event'], 'event_id')
+                if event.state == "draft":
+                    raise ForbiddenException({'source': ''}, 'Co-organizer access is required.')
```
Comment (shubham-padia) [inline app/api/speakers.py]: 404 should be raised, user shouldnt know that event exists

File: app/api/speakers.py [hunk 5]
```diff
@@ -111,42 +92,68 @@ def before_post(self, args, kwargs, data):
...
+
+        if 'sessions' in data:
+            session_ids = data['sessions']
+            for session_id in session_ids:
+                if not has_access('is_session_self_submitted', session_id=session_id):
+                    raise ForbiddenException({'source': ''}, 'Session must be self submitted.')
```
Comment (shubham-padia) [inline app/api/speakers.py]: same here, 404 should be raised

File: app/api/speakers.py [hunk 6]
```diff
@@ -111,42 +92,68 @@ def before_post(self, args, kwargs, data):
         :return:
         """
         require_relationship(['event', 'user'], data)
-        kwargs['user_id'] = current_identity.id
 
-    def before_create_object(self, data, view_kwargs):
+        if not has_access('is_coorganizer', event_id=data['event']):
+            if has_access('auth_required'):
```
Comment (shubham-padia) [inline app/api/speakers.py]: I don't think ` if has_access('auth_required'): ` is required here as `is_coorganizer` already has jwt_required decorator and will check whether user is logged in or not

**Comment by enigmaeth** [comment]
@shubham-padia All issues resolved. Please check.

**Pull request merged**